### ISSUE TITLE
Map support

### ISSUE THREAD

**Issue by asdine** [asdine/storm#103]
Adding map support to Storm could allow users to store dynamic values and to be able to query them as if they were saved with a structure.

The main differences when saving with a map are:
- there is no name
- there are no tags (index, increment, etc.)

So the `Save` method cannot be used as it is now.

The `Set` method though is more flexible, it takes anything and we can pass the bucket name.

```go
Set(bucketName string, k, v interface{}) error
```

We can already do this:

```go

value := map[string]interface{}{
  "name": "John",
  "group": "Staff",
  "Age": 10,
  "UUID": "aaa-bbb-ccc",
}

db.Set("mybucket", 10, value)
```

We could imagine adding options to `Set` to specify indexes and more

```go
db.Set("mybucket", 10, value, storm.Index("group"), storm.Unique("UUID"), storm.Increment("Age"))
```

But specifying this at every call can be boring.
So maybe we can use `Init`, to initialize the bucket before using it

```go
db.Init("mybucket", storm.Index("group"), storm.Unique("UUID"), storm.Increment("Age"))
```

and then

```go
db.Set("mybucket", 10, value)
```

**Comment by bep** [comment]
>But specifying this at every call can be boring.

I wouldn't say that, and it is much better than adding this to some global setting that you forgot you set ... and maybe do not want for all of your maps.

Any sensible programmer would do this in one place:

```go
myCommonBucketOptions := {storm.Index("group"), storm.Unique("UUID"), storm.Increment("Age"))}

...

// Then reuse those same options when saving a map of a "given type"
db.Save(map, myCommonBucketOptions)
```

**Comment by asdine** [comment]
You're right, the only problem i see here is that using `Save` wouldn't work as the bucket name is still missing.
`db.Set()` seems like a good fit for maps

**Comment by asdine** [comment]
The other issue is that we are limited to `map[string]interface{}`. It's the only data structure that matches the current API and that is the map equivalent of a struct.

**Comment by leeola** [comment]
I just ran into this issue. I've got the need to support indexing of dynamic data structures, of which can only be mapped to `map[string]interface{}`, and had *(incorrectly it seems)* assumed Storm was able to index.

Is there no way to index this dynamic data structures via Storm currently? I'd be happy with less than elegant solution as an interim. Thoughts?

So far i'm not seeing anything hopeful based on the godoc.

**Comment by leeola** [comment]
I'm poking around through the source and it appears like much of the rejection of map's are due to automatically building [the configs](https://github.com/asdine/storm/blob/master/extract.go#L23-L39).

As a manual interim in a fork, would it theoretically work to supply my own configs? Eg, expose the configs to public, add a new method `SaveConfig(StructConfig)`, and that's all? Seems easy enough, but i would appreciate feedback.

Note that i don't expect this to be merged, i'm trying to hack in support since this issue might take a bit, as it has no milestone atm :)

**Comment by leeola** [comment]
So i hacked Storm to accept the config, and it appears to work, however i think that method is impractical. The reason being that Storm uses struct references so frequently that not having structs seems impractical, at least with only modifying a single value.

Furthermore, i'm not interested in storing the data, just indexing it. Guess as an interim i'll find another DB or attempt to implement my own indexer.

Looking forward to progress here, thanks!

**Comment by asdine** [comment]
This issue requires a lot of work because for now Storm is too dependant on Structs.

The only solution i see it to work with `Set` and `Get` and make it possible to pass some options to specify what to index. (you can see some examples above)
`Set` and `Get` can store any data.

Another solution would be to specialise and create a `SaveMap` and `GetMap()` which would be easier to implement and would have full support for dynamic indexes.
